### PR TITLE
feat(api): underscore convenience api

### PR DESCRIPTION
**Pull request by cpcloud**
This PR is an experiment to play with a way of avoiding `lambda`s using `_`.

Here's an example:

```python
t = ibis.table(dict(a="int64"), name="t")

expr = (
    t.mutate(b=_.a - 2)
    .select([(_.a.sum() + _.b.count()).name("b")])
    .group_by(key=_.b.cast("string"))
    .aggregate(total=_.b.mean())
)
```

The way this works is that a `Deferred` object graph is built, and resolved
element by element as needed, for example, in a projection or aggregation.
Every `Deferred` instance has a `resolve` attribute that accepts an ibis
expression and returns another ibis expression effectively replaying the graph.
The `_` object's `resolve` attribute is the identity function, which resolves to the
table expression from which the rest of the graph is evaluated over.

### CODE REVIEW

**Comment by kszucs** [comment]
Pretty nice and concise implementation! Going to have a more thorough look tomorrow.

File: ibis/expr/deferred.py
Comment (kszucs) [inline ibis/expr/deferred.py:34]: Also `__getitem__`?

File: ibis/expr/deferred.py
Comment (kszucs) [inline ibis/expr/deferred.py:115]: Is this for making it pluggable? If not could we avoid using singledispatch?

File: ibis/expr/deferred.py
Comment (kszucs) [inline ibis/expr/deferred.py:13]: We can use `Annotable` here as well.

File: ibis/expr/types/relations.py
Comment (kszucs) [inline ibis/expr/types/relations.py:175]: I tend to prefer more explicit branches using `elif` and `else`.

File: ibis/expr/deferred.py
Comment (kszucs) [inline ibis/expr/deferred.py:24]: Wouldn't it be better to define a local function for better readability?

Comment (kszucs) [reply ibis/expr/deferred.py:24]: Something like: 

```python
def resolve(expr):
    expr = self.result(expr)
    args = (_resolve(arg, expr=expr) for arg in args)
    kwargs = {name: _resolve(arg, expr=expr) for name, arg in kwargs.items()}
    return func(expr, *args, **kwargs)
```

File: ibis/expr/deferred.py
Comment (cpcloud) [reply ibis/expr/deferred.py:13]: I'm not sure we get much by doing that here, and it looks to require use of `rlz` which introduces a circular import.

Comment (kszucs) [reply ibis/expr/deferred.py:13]: Indeed, perhaps validate that `resolve` is callable then?

File: ibis/expr/types/relations.py
Comment (cpcloud) [reply ibis/expr/types/relations.py:175]: Done.

File: ibis/expr/deferred.py
Comment (cpcloud) [reply ibis/expr/deferred.py:24]: Roger, done.

File: ibis/expr/deferred.py
Comment (cpcloud) [reply ibis/expr/deferred.py]: Removed.

File: ibis/expr/deferred.py
Comment (cpcloud) [reply ibis/expr/deferred.py:33]: Added.

File: ibis/expr/deferred.py
Comment (cpcloud) [reply ibis/expr/deferred.py:12]: `resolve` should always be callable, and isn't user facing, so I'll add an `assert`.

Comment (cpcloud) [reply ibis/expr/deferred.py:12]: Done.

**Review by kszucs — APPROVED** [review]

**Comment by cpcloud** [comment]
Since we're releasing today, let's wait to merge this until after that.

**Comment by cpcloud** [comment]
cc @machow, I took inspiration from https://github.com/machow/siuba here 😄

**Comment by machow** [comment]
Hey--glad it was handy inspiration! I was tinkering with ibis, and was using lambdas a lot--so this format would go a long way (for a least the way R people are used to :). cc @pwwang, who also is probably interested in libraries doing this kind of thing..

**Comment by kszucs** [comment]
@cpcloud can we go ahead and merge this?

**Comment by cpcloud** [comment]
Yep. I was waiting to see if any more bugs from the 3.0 release shook out before merging any feature PRs, but I think it's probably fine to start merging these. We'll likely release a 3.1 in a month or two anyway so that folks can pick up non-breaking changes along the way to 4.0.

**Pull request merged**
